<a href="https://colab.research.google.com/github/PosgradoMNA/actividades-de-aprendizaje-jesusrincon1328/blob/main/Data%20Analysis%20with%20Python/Module_2__Data_Wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Data Analysis with Python**

**Alumno:** Jesús Angel Rincón Ruiz

**Matricula:** A01793960

**Módulo 2 - Data Wrangling**

Objetivos de aprendizaje:

*   Manejo de falta de valores
*   Correccion de formato
*   Estandarizar y normalizar los datos


# **Apuntes y laboratorio**

**Cual es el proposito de Data Wrangling?**

Data Wrangling es el proceso de convertir los datos de su formato inicial a un formato que podria ser mejor para su analisis.



In [2]:
import pandas as pd
import matplotlib.pylab as plt
import numpy as np

In [3]:
filename =  "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv"


In [4]:
#creamos una lista que sera utilizada como encabezado de las columnas
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

In [5]:
df = pd.read_csv(filename, names = headers)
df.head() #para visualizar las primeras 5 filas

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [6]:
#los signos de interrogación que aparecieron son datos faltantes que perjudicaran el analisis posterior

**Pasos para trabajar con datos faltantes:**

*   Identificar los datos faltantes
*   Manejar datos faltantes
*   Corregir el formato de los datos

# **Identificación y manejo de datos faltantes**

**Identificar datos faltantes**

Convertir "?" to NaN

In [7]:
#utilizaremos la siguiente función .replace(A, B, inplace = True) 
#para reemplazar A por B

#reemplzar "?" por NaN (Not a Number)

In [8]:
df.replace("?", np.nan, inplace = True)
df.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

Hay dos metodos que podemos utilizar para identificar los datos faltantes:

*   **.isnull()**
*   **.notnull()**

La salida es un valor booleano, "True" significa que es un dato faltantes mientras que "False" no lo es.





In [9]:
missing_data =  df.isnull()
missing_data.head(5)

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0       False         False            False       False  ...        False   
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

In [10]:
#contar datos faltantes en cada columna

In [11]:
for columns in missing_data.columns.values.tolist():
    print(missing_data[columns].value_counts())
    print(columns)
    print("")

False    205
Name: symboling, dtype: int64
symboling

False    164
True      41
Name: normalized-losses, dtype: int64
normalized-losses

False    205
Name: make, dtype: int64
make

False    205
Name: fuel-type, dtype: int64
fuel-type

False    205
Name: aspiration, dtype: int64
aspiration

False    203
True       2
Name: num-of-doors, dtype: int64
num-of-doors

False    205
Name: body-style, dtype: int64
body-style

False    205
Name: drive-wheels, dtype: int64
drive-wheels

False    205
Name: engine-location, dtype: int64
engine-location

False    205
Name: wheel-base, dtype: int64
wheel-base

False    205
Name: length, dtype: int64
length

False    205
Name: width, dtype: int64
width

False    205
Name: height, dtype: int64
height

False    205
Name: curb-weight, dtype: int64
curb-weight

False    205
Name: engine-type, dtype: int64
engine-type

False    205
Name: num-of-cylinders, dtype: int64
num-of-cylinders

False    205
Name: engine-size, dtype: int64
engine-size

False    205
N

Como resumen de la información anterior tenemos que cada columna tiene 205 filas y 7 de las columnas tienen datos faltantes.

1.  "normalized-losses": 41 missing data
2.  "num-of-doors": 2 missing data
3.  "bore": 4 missing data
4.  "stroke" : 4 missing data
5.  "horsepower": 2 missing data
6.  "peak-rpm": 2 missing data
7.  "price": 4 missing data

**Manejo de datos faltantes**

1.  Eliminar los datos
    * Eliminar la fila completa
    * Eliminar la columna completa
2.  Reemplazar los datos
    * Reemplazar los datos por el promedio
    * Reemplazar los datos por frecuencia
    * Reemplazar los datos basado en otra funcion

In [12]:
#Calcular el promedio de "normalized-losses"
avg_norm_loss = df["normalized-losses"].astype("float").mean(axis=0)
print("Averge of normalized losses:", avg_norm_loss)

Averge of normalized losses: 122.0


In [15]:
#Reemplazar "NaN" con el promedio de la columna "normalized-losses" 
df["normalized-losses"].replace(np.nan, avg_norm_loss, inplace = True)
df.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3             122.0  alfa-romero       gas        std          two   
1          3             122.0  alfa-romero       gas        std          two   
2          1             122.0  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [17]:
#Calcular el promedio de la columna "bore"
avg_bore = df["bore"].astype("float").mean(axis=0)
print("Average of bore:", avg_bore )

Average of bore: 3.3297512437810943


In [18]:
#reemplazar "NaN" con el promedio de la columna "bore"
df["bore"].replace(np.nan, avg_bore, inplace = True)
df.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3             122.0  alfa-romero       gas        std          two   
1          3             122.0  alfa-romero       gas        std          two   
2          1             122.0  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [19]:
#reemplazar "NaN" en columna "stroke" con el promedio

avg_stroke = df["stroke"].astype("float").mean(axis = 0)
print("Average of stroke:", avg_stroke)
df["stroke"].replace(np.nan, avg_stroke, inplace = True)

Average of stroke: 3.255422885572139


In [20]:
#reemplazar "NaN" en columna "horsepower" con el promedio

avg_horsepower = df['horsepower'].astype('float').mean(axis=0)
print("Average horsepower:", avg_horsepower)
df['horsepower'].replace(np.nan, avg_horsepower, inplace=True)

Average horsepower: 104.25615763546799


In [21]:
#reemplazar "NaN" en columna "peak-rpm" con el promedio

avg_peakrpm=df['peak-rpm'].astype('float').mean(axis=0)
print("Average peak rpm:", avg_peakrpm)
df['peak-rpm'].replace(np.nan, avg_peakrpm, inplace=True)

Average peak rpm: 5125.369458128079


In [ ]:
#para saber cuales valores estan presentes en una columna podemos utilizar el metodo "value_counts()"

In [22]:
df["num-of-doors"].value_counts()

four    114
two      89
Name: num-of-doors, dtype: int64